src/common/src/util/value_encoding/row_encoding.rs

```diff
@@ -0,0 +1,221 @@
+        match datum_num {
+            Width::Mid(_) => {
+                self.flag |= 0b0100;
+            }
+            Width::Large(_) => {
+                self.flag |= 0b1000;
+            }
+            Width::Extra(_) => {
+                self.flag |= 0b1100;
+            }
+        }
+    }
+
+    fn set_big(&mut self, maybe_offset: Vec<usize>, max_offset: usize) {
+        self.offsets = vec![];
+        match max_offset {
+            n if n <= u8::MAX as usize => {
+                self.flag |= 0b01;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u8(m as u8));
+            }
+            n if n <= u16::MAX as usize => {
+                self.flag |= 0b10;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u16_le(m as u16));
+            }
+            n if n <= u32::MAX as usize => {
+                self.flag |= 0b11;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u32_le(m as u32));
+            }
+            _ => unreachable!("encoding length exceeds u32"),
+        }
+    }
+
+    fn encode(&mut self, datum_refs: impl Iterator<Item = impl ToDatumRef>) {
+        assert!(
+            self.buf.is_empty(),
+            "should not encode one RowEncoding object multiple times."
+        );
+        let mut maybe_offset = vec![];
+        for datum in datum_refs {
+            maybe_offset.push(self.buf.len());
+            if let Some(v) = datum.to_datum_ref() {
+                serialize_scalar(v, &mut self.buf);
+            }
+        }
+        let max_offset = *maybe_offset
+            .last()
+            .expect("should encode at least one column");
+        self.set_big(maybe_offset, max_offset);
+    }
+}
+
+pub struct Serializer {
+    encoded_column_ids: Vec<u8>,
+    datum_num: Width,
+    encoded_datum_num: Vec<u8>,
+}
+
+impl Serializer {
+    pub fn new(column_ids: &[ColumnId]) -> Self {
+        let mut encoded_column_ids = Vec::with_capacity(column_ids.len() * 4);
+        for id in column_ids {
+            encoded_column_ids.put_i32_le(id.get_id());
+        }
+        let datum_num = match column_ids.len() {
+            n if n <= u8::MAX as usize => Width::Mid(n as u8),
+            n if n <= u16::MAX as usize => Width::Large(n as u16),
+            n if n <= u32::MAX as usize => Width::Extra(n as u32),
+            _ => unreachable!("the number of columns exceeds u32"),
+        };
+        let mut encoded_datum_num = vec![];
+        serialize_width(datum_num, &mut encoded_datum_num);
+        Self {
+            encoded_column_ids,
+            datum_num,
+            encoded_datum_num,
+        }
+    }
+
+    pub fn serialize_row_column_aware(&self, row: impl Row) -> Vec<u8> {
+        let mut encoding = RowEncoding::new();
+        encoding.encode(row.iter());
```

Comment:
How about directly `set_width` in the `encode`? BTW, we should assert that `row.len()` equals `self.datum_num` here.